display_list/geometry/dl_region.h

```diff
@@ -0,0 +1,65 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_DISPLAY_LIST_GEOMETRY_REGION_H_
+#define FLUTTER_DISPLAY_LIST_GEOMETRY_REGION_H_
+
+#include "third_party/skia/include/core/SkRect.h"
+
+#include <vector>
+
+namespace flutter {
+
+/// Represents a region as a collection of non-overlapping rectangles.
+/// Implements a subset of SkRegion functionality optimized for quickly
+/// converting set of overlapping rectangles to non-overlapping rectangles.
+class DlRegion {
+ public:
+  /// Creates region by bulk adding the rectangles./// Matches
+  /// SkRegion::op(rect, SkRegion::kUnion_Op) behavior.
+  explicit DlRegion(std::vector<SkIRect>&& rects);
```

Comment:
I'm concerned about the danger of modifying the incoming vector of rects as a side effect. A public API should probably be written as `const std::vector<...>& rects` for safety. That might hurt performance by adding an extra copy before the sorting, but perhaps that can be regained by sorting a vector (or array) of pointers to the rects rather than sorting the vector itself?

Also, a constructor that takes (a constant list of) SkRects and turns them into IRects might be useful for more than just the case of calling it from DlRTree. If it has to live in either place, it might as well live here (and then we avoid an additional copy since we can guarantee that after translation to IRects, the new local vector is safe for modifying in place).

Reply:
I tried sorting the vector of pointers and it did not show any improvement to sort the actual list. I think the extra memmove needed to shuffle around SkIRects and the cost of deferecing SkIRect pointers are not that different.

The constructor can not be called without `std::move`, which I'd say is clear indication that DlRegion takes ownership of the incoming region.

I can add a `const & vector` constructor that copies the list.

Reply:
The fact that you must move it makes sense.